src/Eventso.Subscription/Observing/DeadLetter/PoisonEventHandler.cs

```diff
@@ -0,0 +1,99 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Eventso.Subscription.Observing.DeadLetter
+{
+    public sealed class PoisonEventHandler<TEvent> : IEventHandler<TEvent>
+        where TEvent : IEvent
+    {
+        private const string PredecessorParkedReason = "Predecessor of event is poison and parked.";
+
+        private readonly IPoisonEventInbox<TEvent> _poisonEventInbox;
+        private readonly IDeadLetterQueueScopeFactory _deadLetterQueueScopeFactory;
+        private readonly IEventHandler<TEvent> _inner;
+
+        public PoisonEventHandler(
+            IPoisonEventInbox<TEvent> poisonEventInbox,
+            IDeadLetterQueueScopeFactory deadLetterQueueScopeFactory,
+            IEventHandler<TEvent> inner)
+        {
+            _poisonEventInbox = poisonEventInbox;
+            _deadLetterQueueScopeFactory = deadLetterQueueScopeFactory;
+            _inner = inner;
+        }
+
+        public async Task Handle(TEvent @event, CancellationToken cancellationToken)
+        {
+            if (await _poisonEventInbox.IsStreamPoisoned(@event, cancellationToken))
+            {
+                await _poisonEventInbox.Add(
+                    new [] { new PoisonEvent<TEvent>(@event, PredecessorParkedReason) },
+                    cancellationToken);
+                return;
+            }
+
+            using var dlqScope = _deadLetterQueueScopeFactory.Create(@event);
+
+            await _inner.Handle(@event, cancellationToken);
+
+            var poisonEvents = dlqScope.GetPoisonEvents();
+            if (poisonEvents.Count == 0)
+                return;
+
+            await _poisonEventInbox.Add(poisonEvents, cancellationToken);
+        }
+
+        public async Task Handle(IConvertibleCollection<TEvent> events, CancellationToken token)
+        {
+            var poisonStreamsEvents = await _poisonEventInbox.GetPoisonStreamsEvents(events, token);
+
+            FilterPoisonEvents(events, poisonStreamsEvents, out var withoutPoison, out var poison);
```

Comment:
its not clear for me why do we need poison collection here

Reply:
to separate poison streams in source events collection